### PR TITLE
feat: return discriminated union type from decodeFunctionData

### PR DESCRIPTION
**Pull request by fubhy**
I'm probably missing sth. here as I've not spent that much time with `abitype` yet. 

Btw., I cannot emphasize enough how much you guys are knocking it out of the park here). Test suite, type harness, etc. everything is such a pleasant experience to work with. Refactoring, writing tests, adding features, it's hands down the easiest / lowest barrier of entry experience I've ever had to contributing to a library.

ref: https://github.com/wagmi-dev/viem/discussions/264

### CODE REVIEW

File: src/utils/abi/decodeFunctionData.test.ts [hunk 1]
```diff
@@ -58,43 +58,43 @@ test('bar(uint256)', () => {
...
-          name: 'getVoter',
-          outputs: [],
-          stateMutability: 'nonpayable',
-          type: 'function',
-        },
-      ] as const,
-      data: '0xf37414670000000000000000000000000000000000000000000000000000000000010f2c0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac0000000000000000000000000000000000000000000000000000000000000029',
-    }),
-  ).toEqual({
+  const result = decodeFunctionData({
```
Comment (jxom) [inline src/utils/abi/decodeFunctionData.test.ts:61]: Can we add a test of what the type will look like with more than one ABI item?

Comment (fubhy) [reply src/utils/abi/decodeFunctionData.test.ts:61]: Yep, good idea.

Comment (fubhy) [reply src/utils/abi/decodeFunctionData.test.ts:61]: @jxom Added!

File: src/utils/abi/decodeFunctionData.ts [hunk 2]
```diff
@@ -1,21 +1,37 @@
...
+import type { GetFunctionArgs, Hex } from '../../types'
 import { slice } from '../data'
 import { getFunctionSelector } from '../hash'
 import { decodeAbiParameters } from './decodeAbiParameters'
 import { formatAbiItem } from './formatAbiItem'
 
-export type DecodeFunctionDataParameters = {
-  abi: Abi | readonly unknown[]
+export type DecodeFunctionDataParameters<
+  TAbi extends Abi | readonly unknown[],
```
Comment (jxom) [inline src/utils/abi/decodeFunctionData.ts:11]:
```suggestion
  TAbi extends Abi | readonly unknown[] = Abi,
```

File: src/utils/abi/decodeFunctionData.ts [hunk 3]
```diff
@@ -1,21 +1,37 @@
...
+> = {
+  abi: TAbi
   data: Hex
 }
 
-export function decodeFunctionData({
+export type DecodeFunctionDataReturnType<
+  TAbi extends Abi | readonly unknown[],
```
Comment (jxom) [inline src/utils/abi/decodeFunctionData.ts:18]:
```suggestion
  TAbi extends Abi | readonly unknown[] = Abi,
```

**Review by jxom — APPROVED** [review]
Just need a changeset, and we are good to go!

**Review by jxom — APPROVED** [review]

**Pull request merged**